src/shadowsocks_config.ts

```diff
@@ -338,3 +339,63 @@ export const SIP002_URI = {
     return `ss://${userInfo}@${uriHost}:${port.data}/${queryString}${hash}`;
   },
 };
+
+export interface OnlineConfig {
+  // URL endpoint to retrieve a Shadowsocks configuration.
+  readonly url: string;
+  // Any additional configuration (e.g. `certFp`, `httpMethod`, etc.) may be stored here.
+  readonly extra: {[key: string]: string};
+}
+
+// Ref: https://github.com/shadowsocks/shadowsocks-org/issues/89
+export const SIP008_URI = {
+  PROTOCOL: 'ssconf',
+
+  validateProtocol: (uri: string) => {
+    if (!uri || !uri.startsWith(SIP008_URI.PROTOCOL)) {
+      throw new InvalidUri(`URI must start with "${SIP008_URI.PROTOCOL}"`);
+    }
+  },
+
+  parse: (uri: string): OnlineConfig => {
+    SIP008_URI.validateProtocol(uri);
+
+    // URL parser for expedience, replacing the protocol "ssconf" with "https" to ensure correct
+    // results, otherwise browsers like Safari fail to parse it.
+    const inputForUrlParser = uri.replace(new RegExp(`^${SIP008_URI.PROTOCOL}`), 'https');
+    // The built-in URL parser throws as desired when given URIs with invalid syntax.
+    const urlParserResult = new URL(inputForUrlParser);
+    // Use ValidatedConfigFields subclasses (Host, Port, Tag) to throw on validation failure.
+    const uriFormattedHost = urlParserResult.hostname;
+    let host: Host;
+    try {
+      host = new Host(uriFormattedHost);
+    } catch (_) {
+      // Could be IPv6 host formatted with surrounding brackets, so try stripping first and last
+      // characters. If this throws, give up and let the exception propagate.
+      host = new Host(uriFormattedHost.substring(1, uriFormattedHost.length - 1));
+    }
+    // The default URL parser fails to recognize the default HTTPs port (443).
+    const port = new Port(urlParserResult.port || '443');
+
+    // Parse extra parameters from the tag, which has the format `#key0=val0;key1=val1...[;]`
+    const extra = {} as {[key: string]: string};
+    const tag = new Tag(decodeURIComponent(urlParserResult.hash.substring(1)));
+    // Convert tag to search parameters to leverage URLSearchParams parsing.
+    const params = new URLSearchParams(tag.data.replace(';', '&'));
```

Comment:
Why are we using `;` instead of regular format for search params?
How do you represent a value with `;`?

Reply:
Changed to URL search params format in b3f2f19.